### ISSUE TITLE
Investigate potential zombie child problem

### ISSUE THREAD

**Issue by nvie** [liveblocks/liveblocks#511]
Last week, we published a bugfix release 0.18.2 (see https://github.com/liveblocks/liveblocks/pull/506). The bug targeted by this issue might be an instance of the well-known zombie child problem in React. We should investigate a bit deeper and see if there is a more generic fix for this.

**Comment by nvie** [comment]
🧟 I tried doing a bit more digging on the zombie child problem we found last week, and I think it may only be an issue in React 17. In a project on React 18, I'm **unable to reproduce the bug**. This is because in React 18, all state updates are batched automatically, no matter what the source of the events is.

In React 17, this is not the case. Only state updates originating from DOM events (i.e. mouse click handlers, etc) are getting batched there, but this does not include events triggered by Websocket messages. Those get executed synchronously.

Here's what happens on the timeline for both, as a comparison:

In React 17:
1. A Websocket message is received because someone leaves the room
1. All selector functions get reevaluated
1. Both parent and child components states happen to need re-rendering. (The parent component because a connection ID is removed from the array, and the child because the child now "selects" `undefined`.)
1. Because the Websocket handler isn't a DOM event, these are two individual state updates, and thus two rerenders
1. At this point, it depends which component receives the notification first. If the parent is rerendered first, it's no problem. The child will get unmounted. But if the child rerenders first, this will throw the error that #506 tried to fix.

In React 18 (steps 1-3 are the same, difference starts at step 4):
1. A Websocket message is received because someone leaves the room
1. All selector functions get reevaluated
1. Both parent and child components states happen to need re-rendering. (The parent component because a connection ID is removed from the array, and the child because the child now "selects" `undefined`.)
1. Since _all_ state updates are batched, they will trigger only a single rerender cycle.
1. Therefore, the parent component will _always_ rerender first. React guarantees this. No zombie children.

**Comment by GuillaumeSalles** [comment]
Thank you for the write-up @nvie! It's really useful and confirms what we discuss in our last call.

My understanding is that this issue also impacts storage selectors. React 16 and 17 are still more used than 18.

Npm downloads during the last seven days:
React 18.2.0: 2,702,853
React 17.0.2: 5,952,572
React 16.14.0: 3,668,975

I see several options:
A. We stop supporting < react 18 from now on the latest version
B. We let the users know that some hooks are only available on react 18+ (is it even true?)
C. We find a way to use `unstable_batchedUpdates` (still work in 18) to support all react versions

I prefer option C just because of the number of npm downloads, and because passing `unstable_batchedUpdates` from the `RoomProvider` (no op by default) might not be that big of a hack if it means fixing roughly 80 of react uses.

 What do you think?

**Comment by nvie** [comment]
Correct. Storage selectors are also impacted. I haven’t verified/replicated that explicitly yet, but it follows from the theory for sure. Can try to replicate that bug there in a test today.

I prefer option C too. Feels too early to be dropping support for older React versions.

The thing I hadn’t figured out yet is how to wrap all internal event hub `.notify()` calls (which live in the client pkg scope) in a `unstable_batchedUpdates` (a React construct). But your idea of passing it to the client from the `RoomProvider` is interesting and might be the solution around that. Can definitely give that a try today 👍
